leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java

```diff
@@ -159,18 +158,45 @@ protected Object handleInvocation(Object proxy, Method method, Object[] args) th
 
     private RuntimeException handleDelegateThrewException(
             LeadershipToken leadershipToken, InvocationTargetException exception) throws Exception {
-        if (exception.getCause() instanceof ServiceNotAvailableException
-                || exception.getCause() instanceof NotCurrentLeaderException) {
-            leadershipStateManager.handleNotLeading(leadershipToken, exception.getCause());
+        Throwable cause = exception.getCause();
+
+        if (cause instanceof ServiceNotAvailableException) {
+            handleLeadershipChangeException(leadershipToken, (ServiceNotAvailableException) cause);
         }
+
         // Prevent blocked lock requests from receiving a non-retryable 500 on interrupts
         // in case of a leader election.
-        if (exception.getCause() instanceof InterruptedException
-                && !leadershipCoordinator.isStillCurrentToken(leadershipToken)) {
+        if (cause instanceof InterruptedException && !leadershipCoordinator.isStillCurrentToken(leadershipToken)) {
             throw leadershipCoordinator.notCurrentLeaderException(
                     "received an interrupt due to leader election.", exception.getCause());
```

Comment:
Do we need to reset the interrupt (possibly if we don't throw)?

```suggestion
        if (cause instanceof InterruptedException) {
            Thread.currentThread().interrupt();
            if (!leadershipCoordinator.isStillCurrentToken(leadershipToken)) {
                throw leadershipCoordinator.notCurrentLeaderException(
                        "received an interrupt due to leader election.", cause);
```

Reply:
I think in this case, we expect the interrupted exception not to come from process shutdown, but rather us _purposefully_ calling interrupt on these thread(s) to prevent them from performing actions while they're not leader. Although, this could lead to a weird case of we're interrupted due to process shutdown and also no longer leader, which would result in odd behavior. 
@jeremyk-91 thoughts?

Reply:
So some of the methods of the underlying classes can throw InterruptedException in the V1 lock service. In that model, we had synchronous blocking of worker threads as they attempted to wait their turn to acquire a lock (yep, it's pretty legacy 😬). So if a node has lost leadership we want to free any of these waiting threads and tell the users that they should get a NCLE, hence the code as written: don't stop the worker thread, but fail the user request with the exception. Our interruption comes from `LockServiceImpl::close`. 

In this case, there is a valid point that if the interrupt originates from process shutdown and not our specific mechanism, then we don't propagate the interrupt flag - though I imagine we have a bounded wait for this after which we re-check the state and/or terminate the process more forcefully. (At least, I haven't seen any issues from this, as far as I can tell.) Not sure I have a good way out of this beyond changing the API of the lock service to use a different mechanism to handle cancellations on closure...

For the non-NCLE/rethrowing case, I'm pretty sure we should reset the flag though.

Reply:
For the async case, our own interrupt flag would never have been set. Seeing an InterruptedException in a computation we kicked off that doesn't seem to relate to leader election, and us still being the leader, yeah I think we should propagate too.